comet/comet_test.go

```diff
@@ -0,0 +1,347 @@
+		Header: &monomer.Header{},
+	}
+	blockStore.AddBlock(headBlock)
+	headCometBlock := headBlock.ToCometLikeBlock()
+	startBlock := &bfttypes.Block{
+		Header: bfttypes.Header{
+			AppHash: []byte{},
+			Height:  1,
+			Time:    time.Now(),
+		},
+	}
+	statusAPI := comet.NewStatus(blockStore, startBlock)
+	result, err := statusAPI.Status(nil)
+	require.NoError(t, err)
+	require.Equal(t, &rpctypes.ResultStatus{
+		NodeInfo: p2p.DefaultNodeInfo{
+			DefaultNodeID:   "",
+			ListenAddr:      "",
+			Network:         headCometBlock.ChainID,
+			Version:         version.TMCoreSemVer,
+			Channels:        []byte("0123456789"),
+			Moniker:         "monomer",
+			ProtocolVersion: p2p.NewProtocolVersion(version.P2PProtocol, version.BlockProtocol, 0),
+		},
+		// We need SyncInfo so the CosmJS tmClient doesn't complain.
+		SyncInfo: rpctypes.SyncInfo{
+			LatestBlockHash:   headCometBlock.Hash(),
+			LatestAppHash:     headCometBlock.AppHash,
+			LatestBlockHeight: headCometBlock.Height,
+			LatestBlockTime:   headCometBlock.Time,
+
+			EarliestBlockHash:   startBlock.Hash(),
+			EarliestAppHash:     startBlock.AppHash,
+			EarliestBlockHeight: startBlock.Height,
+			EarliestBlockTime:   startBlock.Time,
+
+			CatchingUp: false,
+		},
+		ValidatorInfo: rpctypes.ValidatorInfo{},
+	}, result)
+}
+
+func TestBroadcastTx(t *testing.T) {
+	chainID := "0"
+	app := testapp.NewTest(t, chainID)
+	mpool := mempool.New(testutil.NewMemDB(t))
+	broadcastAPI := comet.NewBroadcastTx(app, mpool)
+	tx := testapp.ToTx(t, "k1", "v1")
+	result, err := broadcastAPI.BroadcastTx(nil, tx)
+	require.NoError(t, err)
+	// We trust that the other fields are set correctly.
+	require.Equal(t, uint32(0), result.Code)
+	require.Equal(t, bfttypes.Tx(tx).Hash(), result.Hash.Bytes())
+	// NOTE: I don't think there is a way to check if CheckTx has been run.
+	// Both CheckTxType_New and CheckTxType_Recheck succeed.
+	got, err := mpool.Dequeue()
+	require.NoError(t, err)
+	require.Equal(t, bfttypes.Tx(tx), got)
+}
+
+type mockWSConnection struct {
+	ctx    context.Context
+	t      *testing.T
+	writes chan<- jsonrpctypes.RPCResponse
+}
+
+func newMockWSConnection(t *testing.T, writes chan<- jsonrpctypes.RPCResponse) *mockWSConnection {
+	return &mockWSConnection{
+		ctx:    context.Background(),
+		t:      t,
+		writes: writes,
+	}
+}
+
+func (c *mockWSConnection) GetRemoteAddr() string {
+	return "ws://127.0.0.1:8888/ws"
+}
+
+func (c *mockWSConnection) WriteRPCResponse(_ context.Context, resp jsonrpctypes.RPCResponse) error {
+	if c.writes != nil {
+		c.writes <- resp
+	}
+	return nil
+}
+
+func (c *mockWSConnection) TryWriteRPCResponse(resp jsonrpctypes.RPCResponse) bool {
+	// We know this is only called when writing an error string.
+	var errStr string
+	require.NoError(c.t, json.Unmarshal(resp.Result, &errStr))
+	require.Contains(c.t, errStr, "subscription was canceled (reason")
+	return true
+}
+
+func (c *mockWSConnection) Context() context.Context {
+	return c.ctx
+}
```

Comment:
Move to `testutil/utils.go` ?

Reply:
Will consider if it's used outside of these tests.